lib/trie/hash.go

```diff
@@ -5,195 +5,285 @@ package trie
-		return []byte{0}, nil
+		buffer.Write([]byte{0})
+		return nil
+	default:
+		return fmt.Errorf("%w: %T", ErrNodeTypeUnsupported, n)
 	}
-
-	return nil, nil
 }
 
 func encodeAndHash(n node) ([]byte, error) {
-	h := newHasher(false)
-	defer h.returnToPool()
+	buffer := digestBufferPool.Get().(*bytes.Buffer)
+	buffer.Reset()
+	defer digestBufferPool.Put(buffer)
 
-	encChild, err := h.Hash(n)
+	err := hashNode(n, buffer)
 	if err != nil {
 		return nil, err
 	}
 
-	scEncChild, err := scale.Marshal(encChild)
+	scEncChild, err := scale.Marshal(buffer.Bytes())
 	if err != nil {
 		return nil, err
 	}
 	return scEncChild, nil
 }
 
 // encodeBranch encodes a branch with the encoding specified at the top of this package
-func (h *hasher) encodeBranch(b *branch) ([]byte, error) {
+// to the buffer given.
+func encodeBranch(b *branch, buffer *bytes.Buffer, parallel bool) (err error) {
 	if !b.dirty && b.encoding != nil {
-		return b.encoding, nil
+		_, err = buffer.Write(b.encoding)
+		return err
 	}
-	h.tmp.Reset()
 
 	encoding, err := b.header()
-	h.tmp.Write(encoding)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	h.tmp.Write(nibblesToKeyLE(b.key))
-	h.tmp.Write(common.Uint16ToBytes(b.childrenBitmap()))
+	buffer.Write(encoding)
+	buffer.Write(nibblesToKeyLE(b.key))
+	buffer.Write(common.Uint16ToBytes(b.childrenBitmap()))
 
 	if b.value != nil {
 		bytes, err := scale.Marshal(b.value)
 		if err != nil {
-			return nil, err
-		}
-		h.tmp.Write(bytes)
-	}
-
-	if h.parallel {
-		wg, _ := errgroup.WithContext(context.Background())
-		resBuff := make([][]byte, 16)
-		for i := 0; i < 16; i++ {
-			func(i int) {
-				wg.Go(func() error {
-					child := b.children[i]
-					if child == nil {
-						return nil
-					}
-
-					var err error
-					resBuff[i], err = encodeAndHash(child)
-					if err != nil {
-						return err
-					}
-					return nil
-				})
-			}(i)
-		}
-		if err := wg.Wait(); err != nil {
-			return nil, err
+			return err
 		}
+		buffer.Write(bytes)
+	}
+
+	if parallel {
+		return encodeChildsInParallel(b.children, buffer)
+	}
+	return encodeChildsSequentially(b.children, buffer)
+}
 
-		for _, v := range resBuff {
-			if v != nil {
-				h.tmp.Write(v)
+func encodeChildsInParallel(children [16]node, buffer *bytes.Buffer) (err error) {
+	type result struct {
+		index  int
+		buffer *bytes.Buffer
+		err    error
+	}
+
+	resultsCh := make(chan result)
+
+	for i, child := range children {
+		go func(index int, child node) {
+			buffer := encodingBufferPool.Get().(*bytes.Buffer)
+			buffer.Reset()
+			// buffer is put back in the pool after processing its
+			// data in the select block below.
+
+			err := encodeChild(child, buffer)
+
+			resultsCh <- result{
+				index:  index,
+				buffer: buffer,
+				err:    err,
 			}
+		}(i, child)
+	}
+
+	currentIndex := 0
+	resultBuffers := make([]*bytes.Buffer, len(children))
+	for range children {
+		result := <-resultsCh
+		if result.err != nil && err == nil { // only set the first error we get
```

Comment:
is there ever a case `err` would not be nil here?

Reply:
Yes once it's set by the first encountered `result.err` from the encoding of one of the 16 children.